projects/ngx-charts-on-fhir/src/lib/data-layer/data-layer-manager.service.ts

```diff
@@ -0,0 +1,137 @@
+import { forwardRef, Inject, Injectable } from '@angular/core';
+import { BehaviorSubject, map, merge, Observable, ReplaySubject } from 'rxjs';
+import { DataLayer, DataLayerCollection, ManagedDataLayer } from './data-layer';
+import { DataLayerColorService } from './data-layer-color.service';
+import { DataLayerMergeService } from './data-layer-merge.service';
+import produce, { castDraft } from 'immer';
+import { DataLayerModule } from './data-layer.module';
+
+/** A service for asynchronously retrieving DataLayers */
+export abstract class DataLayerService {
+  abstract name: string;
+  abstract retrieve: () => Observable<DataLayer>;
+}
+
+type DataLayerManagerState = {
+  layers: DataLayerCollection;
+  selected: string[];
+};
+
+const initialState: DataLayerManagerState = {
+  layers: {},
+  selected: [],
+};
+
+/**
+ * A service that retrieves [DataLayer]s from all registered [DataLayerService]s
+ * and provides methods for selecting and ordering the layers.
+ */
+@Injectable({
+  providedIn: forwardRef(() => DataLayerModule),
+})
+export class DataLayerManagerService {
+  constructor(
+    @Inject(DataLayerService) readonly dataLayerServices: DataLayerService[],
+    private colorService: DataLayerColorService,
+    private mergeService: DataLayerMergeService
+  ) {}
+
+  private stateSubject = new BehaviorSubject<DataLayerManagerState>(initialState);
+  private get state() {
+    return this.stateSubject.value;
+  }
+  private set state(value: DataLayerManagerState) {
+    this.stateSubject.next(value);
+  }
+  allLayers$ = this.stateSubject.pipe(map(({ layers }) => Object.values(layers)));
+  selectedLayers$ = this.stateSubject.pipe(map(({ layers, selected }) => selected.map((id) => layers[id])));
+  availableLayers$ = this.allLayers$.pipe(map((layers) => layers.filter((layer) => !layer.selected)));
+
+  // todo: roll this into stateSubject?
+  private timelineRangeSubject = new ReplaySubject<{ min: number; max: number }>();
+  timelineRange$ = this.timelineRangeSubject.asObservable();
+
+  loading = false;
+
+  retrieveAll() {
+    this.loading = true;
+    merge(...this.dataLayerServices.map((service) => service.retrieve())).subscribe({
```

Comment:
@semicolin, can you please add comment around this, how it exactly performs

Reply:
added a comment